doc/api/esm.md

```diff
@@ -241,10 +241,12 @@ There are four types of specifiers:
 Bare specifiers, and the bare specifier portion of deep import specifiers, are
 strings; but everything else in a specifier is a URL.
 
-Only `file://` URLs are supported. A specifier like
+Only `file:` and `data:` URLs are supported. A specifier like
 `'https://example.com/app.js'` may be supported by browsers but it is not
 supported in Node.js.
 
+`data:` URLs only resolve _Bare specifiers_ for builtin modules and _Absolute specifiers_. Resolving _Relative specifiers_ will not work because `data:` is not a [special scheme]. For example, attempting to load `./foo` from `data:text/javascript,import "./foo";` will fail to resolve since there is no concept of relative resolution for `data:` URLs.
```

Comment:
i still don’t understand why it couldn’t be relative to the data-url-importing file.

Reply:
If 2 different files import the same absolute data URL it wouldn't make sense; you would either leak one location as the resolution location (potentially a security issue/confusing) or need to produce multiple modules for the single absolute data URL (doesn't make sense). Also data URLs are inherently a location on their own.

This matches the web

Reply:
If 2 different files import the same absolute data URL it wouldn't make sense; you would either leak one location as the resolution location (potentially a security issue/confusing) or need to produce multiple modules for the single absolute data URL (doesn't make sense). Also data URLs are inherently a location on their own.

This matches the web

Reply:
It makes sense to me; whoever's constructing the string decides how portable it is.

Reply:
@ljharb can you clarify that last comment.

Reply:
Just like URLs (and thus dynamic import specifiers) are relative to the page/module, I’d expect that a data uri would be too. I don’t see why a data uri would need to be portable across modules.

Reply:
@ljharb these are absolute URLs, they aren't relative.

Reply:
I’m referring to the “./foo” in the paragraph we’re commenting on - is that not relative?

Reply:
data uris are absolute.

Reply:
@ljharb within the "data:" scheme it isn't because resolving relative URL strings against a base only happens for [special schemes](https://url.spec.whatwg.org/#special-scheme).

Reply:
I get that’s how the web has decided it should work, but node isn’t the web and can make better choices.

Reply:
@ljharb Sure, but is allowing `import './???'` from within a data url really *better*? What would go in there? Everything after `:`? Is that really an improvement over just writing `data:<full thing>`?

Reply:
@ljharb thats how URL resolution works in general? The web spec is just where the standard URL resolution lives... we don't have a real way of saying that we are changing how URL resolution works here.

Reply:
> I get that’s how the web has decided it should work, but node isn’t the web and can make better choices.

I strongly disapprove of this sentiment in general. If you don’t like the standards, work to change the standards. Just because Node can ignore how the web does things doesn’t mean it should, even if we think it’s better.

Reply:
regardless of how the spec is approached, there isn't really any logical way to define `x:` + `./y` going to anything other than another `x:`.

Reply:
@GeoffreyBooth the standards for the web are *not* the standards for node. node isn't a browser. node can't ignore how the language does things, but in no way need it follow how browsers do things.

My pushback here *is* working to change node's standards.

Reply:
@jkrems it's already allowed, i think, it just won't be relative to the current file; what makes it better to me is that when i've literally typed out a data URL that contains `import('./foo')`, it will be the same as if i typed `import('./foo')` outside the data URL in the same file.

Reply:
@ljharb So your proposal is that we somehow keep track of the referrer of the data URL and the resolve imports using *that* referrer instead of the actual one (the `data:` URL)? That seems like super surprising behavior.

Reply:
It would also mean that if two modules run `data:export default import('./x');`, they would randomly get x resolved based on one of their URLs, depending on execution order. This feels like we're opening a huge can of worms by adding it.

Reply:
@jkrems referrer? i'm suggesting that the URL of the module that imports the data URL should be the base URL for execution of the "data URL module".

Reply:
the data uri is the base url though. the fact that the source happens to be within the url is completely transparent to url resolution.

Reply:
@ljharb as stated above it doesn't make sense even if we make it relative.

if 2 different files with different resolution namespaces a and b `import 'data:text/javascript,import "c";`

```
/a
/a/node_modules/c
/b
/b/node_modules/c
```

We would need to resolve `c` to 2 different values but we only have 1 absolute URL for the data URL. It just doesn't make sense. I will not move forward with such confusion and breakage with how URLs work regardless due to this.

Being an absolute URL, these are not contextual strings. That should not be up for discussion.

Reply:
Why doesn't that makes sense to you? It makes perfect sense to me since that's *exactly how `import()` already operates*.

Reply:
@ljharb There's two different modules here and `import` doesn't usually work the way you describe it.

```js
// module 1 at file:///foo/a.mjs
import '<absolute url>';

// module 2 at <absolute url>
import './x.mjs';
```

You say that iff `<absolute url>` happens to be a `data:` url, then `import './x.mjs` should be resolved relative to module 1. Which isn't how dynamic import or any other import works right now.

Reply:
For one thing, it would be impossible without breaking this invariant

> This operation must be idempotent if it completes normally. Each time it is called with a specific referencingScriptOrModule, specifier pair as arguments it must return the same Module Record instance.

Reply:
@devsnek it wouldn’t; because the referencing module is/should be different each time, based on source location.

@jkrems i see your argument, but i do think that the location of the source text of module 2 matters - it also matters with static imports (because it changes the url or the file location). Making the source text evaluate the same no matter where it originates is what seems strange and inconsistent to me, since that’s not the case otherwise.

Reply:
@ljharb the location of the source text is the same for `a` and `b` in my screenshot though

Reply:
@ljharb this is just acting like how importing "file://c:/foo/lib.mjs" from "file://c:/app/main.mjs" doesn't make foo resolve it's imports against main.mjs . This behavior is not up for change.

Reply:
That's how browsers handle this:

```js
import(`data:text/javascript,export * from "/x.js"`);
```

> Uncaught (in promise) TypeError: Failed to resolve module specifier "/x.js". Invalid relative url or base scheme isn't hierarchical.

Reply:
Note here that resolution has no concept beyond the referring module `data:…`.

Also worth note is that the same data URL only instantiates once:

```js
import(`data:text/javascript,console.log(1)`)
```

The above logs on the first time only.

But something as simple as arguably not intentional differences like normalization of `\0` characters can lead to undesired behaviour if multiple instances was the intent.

Using `data:` comes with those assumed risks imho if you are relying on them, and maybe we want to consider consistency here a more responsible course to take.

Reply:
@devsnek no, they're two different ones. if you stored it in a variable and `import(a)` twice, then it'd be the same for the string, and it's fair to note that it might be a bit strange to say that the location of the `import()` matters as opposed to the string (but also might not).

@bmeck correct, but the *location of foo/lib* does determine that. I'm saying that a data URL has a distinct "location" wherever it's used. (Also statements like "this is not up for change" really don't promote productive discussion; this is a pull *request* and everything about it is up for change).

@SMotaal i realize that in browsers, they've chosen to make a data URI be its own key and for it to have no location. I'm suggesting that that behavior isn't ideal, and that node should make a better choice. (Regarding your first point, if `/x.js` existed as a URL from the root i assume that it would resolve, regardless of where the data URL was imported from)

Reply:
@ljharb that is not the `data:` scheme. You can PR a different scheme with those new behaviors that would break `data:`. This topic is resolved as not a bug. Please open a different PR with a new scheme if you want that behavior

Reply:
I don’t consider it resolved, because “bugs” aren’t the only things discussed on PRs. Please don’t be so dismissive.

If this isn’t how data: can work, then I’d prefer node not add support for data uris at all - why would we want to add a new source of eval?

Reply:
As a user, I expect code that I write that can run in both Node and browsers to run the same way in both Node and browsers.

I agree with @bmeck. If Node adds support for `data:`, it needs to behave the same as how browsers work. If @ljharb wants a “relative `data:`”, that should be proposed as something different, with a new name. It would be very counterintuitive to users for Node’s `data:` to follow divergent semantics to browsers’ `data:`, no matter how much better @ljharb thinks he can improve upon what browsers have shipped.

Reply:
@GeoffreyBooth then i hope node’s planning to remove fs and http access, since browsers don’t have that - if not, then your expectation simply can never be met, and node shouldn’t be making decisions with that expectation in mind.

Reply:
@ljharb I am fine making a new scheme with different behavior from `data:` in a different PR. However, implementing non `data:` compliant URLs should not use the `data:` scheme and thus is not up for issue in this PR. If you do not want to have `data:` URLs at all that is a topic that you can discuss. Changing behavior of what the officially standardized IANA/IETF/WHATWG `data:` scheme does is outside of the scope of this PR.

Reply:
@bmeck fair enough, then let’s discuss that.

Reply:
> As a user, I expect code that I write that can run in both Node and browsers to run the same way in both Node and browsers.

`console.log` behaves differently between platforms (in chrome it prints an inspectable interactive expandable object where possible, in `node` the object is just printed in full), but exists on both. Is that a bug? Platforms _do have_ platform differences - you need to enumerate _exactly_ what you expect to behave the same and why to make a compelling argument, IMO.

Reply:
@weswigham fwiw node implements the same console specification that browsers do and is a conformant implementation.

Reply:
I think the browser precedent is important but I don't think it's necessary to discuss it here. It feels kind of off-topic. What matters here is: Should module specifiers be resolved to the URL of the module they appear in? I feel fairly strongly that it should. I don't want us to create another `eval` and as myself and others have pointed out above, nobody has yet provided semantics for this that aren't actively dangerous.

To make my example from above more explicit, it would be something like this:

```js
// file:///foo/a.mjs
import data from 'data:text/javascript;base64,aW1wb3J0IHggZnJvbSAnLi9hLm1qcyc7DQpleHBvcnQgZGVmYXVsdCB7IHgsIHVybDogaW1wb3J0Lm1ldGEudXJsIH07';
export default 'foo';
console.log(data);

// file:///bar/a.mjs
import data from 'data:text/javascript;base64,aW1wb3J0IHggZnJvbSAnLi9hLm1qcyc7DQpleHBvcnQgZGVmYXVsdCB7IHgsIHVybDogaW1wb3J0Lm1ldGEudXJsIH07';
export default 'bar';
console.log(data);

// data:text/javascript;base64,aW1wb3J0IHggZnJvbSAnLi9hLm1qcyc7DQpleHBvcnQgZGVmYXVsdCB7IHgsIHVybDogaW1wb3J0Lm1ldGEudXJsIH07
import x from './a.mjs';
export default { x, url: import.meta.url };
```

This is a set of modules with only static import where, with @ljharb's semantics, the graph would be order-dependent. I really don't want to introduce that kind of behavior into our module system. To echo others - if somebody wants to introduce it, they should be using something other than `data:` URLs to express it. Maybe not even a URL scheme because it doesn't seem to match any other use of URLs.

Reply:
"data" url semantics have been specified since the late 90's, i don't think we're going to be changing them here.

Reply:
@jkrems i think I’ve been convinced that *since* data urls work like this everywhere else, they should work like that in node too - altho i don’t think they should work like that anywhere.

However, i don’t see why it’s important or useful to create another form of eval in node; what are the use cases for data uris in node?

Reply:
@ljharb new forms of evaluators keep being added in various means and by various efforts (WASM/worker_threads/AsyncGeneratorFunction/etc.), this one increases overall compatibility with other environments. Many of these evaluators are not attempting to limit API authority and are introduced in various ways that grant them authority. In addition Node's default mode of running gives arbitrary power to any module already so limiting evaluators is unlikely to be useful as you could just write an interpreter in JS to act as one. Different means than limiting the number evaluators are needed to prevent access to APIs. Increased compatibility seems valuable in this light.

Use of `data:` is potentially helpful for a variety of uses but largely the ability to place code into the shared module map is the key utility. In theory it could also be used to write programs that never have `fs` access but I find that unlikely in reality.

I would note, the ability to use `data:` based modules when talking about cross compatibility is likely somewhat limited since it would be based upon using modules that don't use platform specific APIs / use feature detection. It is notable that if browsers move forward with import maps, even platform specific import could be shimmed within the `data:` based modules. Similarly, web focused `data:` modules could use HTTP based imports and a user could write a Loader for Node to handle those however appropriate.

Reply:
One interesting case is loaders. If you want to wrap an original URL in a custom wrapper (e.g. for instrumentation), you can use a `data:` URL in a resolver to model that:

```js
export function resolve(parentResolve, specifier, referrerURL) {
  const target = parentResolve(specifier, referrerURL);
  if (wantsToInstrument(target)) {
    return buildInstrumentationModule(target);
  }
  return target;
}
```

Ignore the strawman resolve hook API. I hope the general idea translates anyhow. :)

Reply:
"Placing things in the module cache" seems useless since the key would be the full data URL - as in, it's not like i can use a data URL to simulate import maps.

@jkrems When would it be more convenient (for the loader author) to use strings-as-code in a data URL rather than referencing a separate file that can be syntax highlighted and linted and tested?

Reply:
@ljharb not everything is based upon pulling in other modules for utility, consider a inserting a `data:` dependency in the module src text to ensure it runs prior to other dependencies or creating a shared space like `data:text/javascript,export default new WeakSet();`. Overall this PR is a compatibility gain which on its own is very valuable.

Reply:
You can do the former by creating a separate file and `import`ing it at the top; you can do the latter also with a separate file - what's the utility in avoiding separate files for separate modules?

I agree it increases the union of features between node and browsers, but that is not always valuable on its own - more often than i think is appreciated, some features of one do not belong in the other.

Reply:
@ljharb

> what's the utility in avoiding separate files for separate modules?

Compatibility

> I agree it increases the union of features between node and browsers, but that is not always valuable on its own - more often than i think is appreciated, some features of one do not belong in the other.

What is the counter argument about how adding this is not valuable / this feature does not belong in Node? Is there some harmful nature that we haven't seen yet in these comments? Even if the result is just appreciation from users, that seems like a good thing.

Reply:
> You can do the former by creating a separate file and importing it at the top; you can do the latter also with a separate file - what's the utility in avoiding separate files for separate modules?

Maybe my bad for not being more explicit in my example. `buildInstrumentationModule` would realistically load a file from disk and then mutate it, that's why it received the target URL as an argument. The file on disk cannot contain the target URL because it's only known at runtime. E.g. it may be the URL of where the memcached client is loaded from. And there could be multiple copies of the instrumentation template for multiple copies of the memcached client across the dependency graph.

Reply:
@jkrems and why is mutating it in-memory to produce an evallable string a better approach than writing it initially as a function that takes arguments and returns what you need?

Reply:
I'm not quite following what you are suggesting. If we can just choose to change the interface of the instrumented module - sure. But that's a completely different scenario. Usually when instrumenting you can neither change the consuming module nor the targeted module. Can you give a concrete example of what you were thinking about?

Reply:
I guess I'm having trouble understanding how you can ever craft a useful dynamic-source module, when it resolves only absolutely, without some kind of `import.meta.resolve` function - how could it ever bring in functionality from relative paths otherwise?

Reply:
Since we are crafting it, we can always generate absolute imports. E.g. if my instrumentation file needs access to others, it can import them by absolute URLs, inserted when constructing the source text. It could be as easy as `src.replace('UTIL_URL', new URL('./util.mjs', import.meta.url).href)`.